File: tests/test_i18n.py
from i2ptorrents.i18n import STRINGS, normalize_language, set_language, t


def test_default_language_is_english() -> None:
    set_language("en")
    assert t("settings") == "Settings"
    assert t("add_torrent") == "Add torrent"


def test_russian_language() -> None:
    set_language("ru")
    assert t("settings") == "Настройки"
    assert t("add_torrent") == "Добавить торрент"
    set_language("en")


def test_normalize_language() -> None:
    assert normalize_language("ru") == "ru"
    assert normalize_language("Русский") == "ru"
    assert normalize_language("en") == "en"
    assert normalize_language("de") == "en"


def test_about_strings() -> None:
    set_language("en")
    assert t("about") == "About"
    assert "0.1.0" in t("subtitle", version="0.1.0")
    assert "BSD-3-Clause" in t("about_body", version="0.1.0", author="Vade", license="BSD-3-Clause")
    set_language("ru")
    assert t("about") == "О программе"
    set_language("en")


def test_translation_catalogs_have_the_same_keys() -> None:
    assert set(STRINGS["en"]) == set(STRINGS["ru"])

File: tests/test_i18n.cpp
#include "i18n.hpp"

#include <QTest>

class I18nTests final : public QObject {
    Q_OBJECT

private slots:
    void defaultLanguageIsEnglish();
    void russianLanguage();
    void normalizeLanguageAliases();
    void aboutStrings();
    void translationCatalogsHaveSameKeys();
};

void I18nTests::defaultLanguageIsEnglish()
{
    i2p::setLanguage(QStringLiteral("en"));
    QCOMPARE(i2p::trKey(QStringLiteral("settings")), QStringLiteral("Settings"));
    QCOMPARE(i2p::trKey(QStringLiteral("add_torrent")), QStringLiteral("Add torrent"));
    QVERIFY(i2p::trKey(QStringLiteral("add_torrent_tip")).contains(QStringLiteral("{shortcut}")));
}

void I18nTests::russianLanguage()
{
    i2p::setLanguage(QStringLiteral("ru"));
    QCOMPARE(i2p::trKey(QStringLiteral("settings")), QStringLiteral("Настройки"));
    QCOMPARE(i2p::trKey(QStringLiteral("add_torrent")), QStringLiteral("Добавить торрент"));
    i2p::setLanguage(QStringLiteral("en"));
}

void I18nTests::normalizeLanguageAliases()
{
    QCOMPARE(i2p::normalizeLanguage(QStringLiteral("ru")), QStringLiteral("ru"));
    QCOMPARE(i2p::normalizeLanguage(QStringLiteral("Русский")), QStringLiteral("ru"));
    QCOMPARE(i2p::normalizeLanguage(QStringLiteral("en")), QStringLiteral("en"));
    QCOMPARE(i2p::normalizeLanguage(QStringLiteral("de")), QStringLiteral("en"));
}

void I18nTests::aboutStrings()
{
    i2p::setLanguage(QStringLiteral("en"));
    QCOMPARE(i2p::trKey(QStringLiteral("about")), QStringLiteral("About"));
    const QString body = i2p::trArgs(QStringLiteral("about_body"),
                                     {{QStringLiteral("version"), QStringLiteral("0.1.0")},
                                      {QStringLiteral("author"), QStringLiteral("Vade")},
                                      {QStringLiteral("license"), QStringLiteral("BSD-3-Clause")}});
    QVERIFY(body.contains(QStringLiteral("BSD-3-Clause")));
    i2p::setLanguage(QStringLiteral("ru"));
    QCOMPARE(i2p::trKey(QStringLiteral("about")), QStringLiteral("О программе"));
    i2p::setLanguage(QStringLiteral("en"));
}

void I18nTests::translationCatalogsHaveSameKeys()
{
    i2p::setLanguage(QStringLiteral("en"));
    const QString enSettings = i2p::trKey(QStringLiteral("settings"));
    i2p::setLanguage(QStringLiteral("ru"));
    const QString ruSettings = i2p::trKey(QStringLiteral("settings"));
    QVERIFY(!enSettings.isEmpty());
    QVERIFY(!ruSettings.isEmpty());
}

int runI18nTests(int argc, char *argv[])
{
    I18nTests suite;
    return QTest::qExec(&suite, argc, argv);
}

#include "test_i18n.moc"
